src/ripple/rpc/handlers/GatewayBalances.cpp

```diff
@@ -0,0 +1,228 @@
+#include <ripple/rpc/impl/LookupLedger.h>
+#include <ripple/app/paths/RippleState.h>
+
+namespace ripple {
+
+// Query:
+// 1) Specify ledger to query.
+// 2) Specify issuer account (cold wallet) in "account" field.
+// 3) Specify accounts that hold gateway assets (such as hot wallets)
+//    using "hotwallet" field which should be either a string (if just
+//    one wallet) or an array of strings (if more than one).
+
+// Response:
+// 1) Array, "obligations", indicating the total obligations of the
+//    gateway in each currency. Obligations to specified hot wallets
+//    are not counted here.
+// 2) Object, "balances", indicating balances in each account
+//    that holds gateway assets. (Those specified in the "hotwallet"
+//    field.)
+// 3) Object of "assets" indicating accounts that owe the gateway.
+//    (Gateways typically do not hold positive balances. This is unusual.)
+
+// gateway_balances [<ledger>] <account> [<howallet> [<hotwallet [...
+
+Json::Value doGatewayBalances (RPC::Context& context)
+{
+    auto& params = context.params;
+
+    // Get the current ledger
+    Ledger::pointer ledger;
+    Json::Value result (RPC::lookupLedger (params, ledger, context.netOps));
+
+    if (!ledger)
+        return result;
+
+    if (!(params.isMember (jss::account) || params.isMember (jss::ident)))
+        return RPC::missing_field_error (jss::account);
+
+    std::string const strIdent (params.isMember (jss::account)
+        ? params[jss::account].asString ()
+        : params[jss::ident].asString ());
+
+    int iIndex = 0;
+
+    if (params.isMember (jss::account_index))
+    {
+        auto const& accountIndex = params[jss::account_index];
+        if (!accountIndex.isUInt() && !accountIndex.isInt ())
+            return RPC::invalid_field_message (jss::account_index);
+        iIndex = accountIndex.asUInt ();
+    }
+
+    bool const bStrict = params.isMember (jss::strict) &&
+            params[jss::strict].asBool ();
+
+    // Get info on account.
+    bool bIndex; // out param
+    RippleAddress naAccount; // out param
+    Json::Value jvAccepted (RPC::accountFromString (
+        ledger, naAccount, bIndex, strIdent, iIndex, bStrict, context.netOps));
+
+    if (!jvAccepted.empty ())
+        return jvAccepted;
+
+    context.loadType = Resource::feeHighBurdenRPC;
+
+    result[jss::account] = naAccount.humanAccountID();
+    auto accountID = naAccount.getAccountID();
+
+    // Parse the specified hotwallet(s), if any
+    std::set <Account> hotWallets;
+
+    if (params.isMember ("hotwallet"))
+    {
+        Json::Value const& hw = params["hotwallet"];
+        bool valid = true;
+
+        auto addHotWallet = [&valid, &hotWallets](Json::Value const& j)
+        {
+            if (j.isString())
+            {
+                RippleAddress ra;
+                if (! ra.setAccountPublic (j.asString ()) &&
+                    ! ra.setAccountID (j.asString()))
+                {
+                    valid = false;
```

Comment:
It might be easier just to have the lambda return a bool - `return false;` here and `return true;` in the one success case.

To taste!